staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/apiserver/pkg/endpoints/metrics"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+
+	"k8s.io/klog/v2"
+)
+
+type filterStartTimestampKeyType int
+
+const filterStartTimestampKey filterStartTimestampKeyType = iota
+
+func WithFiletrStartTimestamp(parent context.Context, startTimestamp time.Time) context.Context {
+	return apirequest.WithValue(parent, filterStartTimestampKey, startTimestamp)
+}
+
+func FilterStartTimestampFrom(ctx context.Context) (time.Time, bool) {
+	info, ok := ctx.Value(filterStartTimestampKey).(time.Time)
+	return info, ok
+}
+
+func DecorateFilter(chain http.Handler, name string, longRunningRequestCheck apirequest.LongRunningRequestCheck, decorate func(h http.Handler) http.Handler) http.Handler {
```

Comment:
I don't understand this file at all, why is it so complicated? Is this for testing?

Reply:
I guess this is for timestamping before/after? I wouldn't use the word "decorate" for that, it's too general.

I'm not sure about this approach in general, P&F might already know the duration. I guess I'd expect to update this metric when P&F makes a decision about the request.